iwyu.cc

```diff
@@ -1380,30 +1370,70 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
-            GetFileEntry(use_loc), GetFileEntry(used_decl)))
-      return GetInstantiationLoc(use_loc);
+  // Return any forward-declare of decl earlier in the same file as loc.
+  const NamedDecl* GetForwardDeclareInSameFile(const Decl* decl,
+                                               SourceLocation loc) {
+    set<const NamedDecl*> redecls = GetClassRedecls(DynCastFrom(decl));
+    for (const NamedDecl* redecl : redecls) {
+      if (IsBeforeInSameFile(redecl, loc)) {
+        return redecl;
+      }
+    }
+
+    return nullptr;
+  }
+
+  // Get the canonical use location for a location,decl pair.
+  // Decide whether the file expanding the macro or the file defining the macro
+  // should be held responsible for a use.
+  SourceLocation GetCanonicalUseLocation(SourceLocation use_loc,
+                                         const NamedDecl* decl) {
+    // If we're not in a macro, just echo the use location.
+    if (!IsInMacro(use_loc))
+      return use_loc;
+
+    // If we're in a macro, and the macro file forward-declares decl, make sure
+    // we keep that forward-declaration.
+    const NamedDecl* fwd_decl = GetForwardDeclareInSameFile(decl, use_loc);
+    if (fwd_decl) {
+      const FileEntry* used_in = GetFileEntry(use_loc);
+      preprocessor_info().FileInfoFor(used_in)->ReportForwardDeclareUse(
+          use_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
+          nullptr);
+    }
+
+    // Resolve the best use location based on our current knowledge.
+    //
+    // 1) If the macro definition file forward-declares the used decl, that's a
+    //    hint that it wants the expansion location to take responsibility.
+    // 2) If the symbol being used is passed as an argument to the macro, we
+    //    want the expansion location to take responsibility (that thing they
+    //    passed in is something they know about, and the macro doesn't.)
+    // 3) If the use_loc is in <scratch space>, we assume it's formed by macro
+    //    argument concatenation, and attribute responsibility to the expansion
+    //    location.
+    //
+    // Otherwise, the macro file is responsible for including the symbol.
+    SourceLocation expansion_loc = GetInstantiationLoc(use_loc);
+    const char* responsible = "expansion";
+
+    if (fwd_decl != nullptr) {
+      VERRS(5) << "Macro file has a forward-decl attribution hint.\n";
+      use_loc = expansion_loc;
+    } else if (IsInScratchSpace(use_loc)) {
+      VERRS(5) << "Decl was used in <scratch space>, presumably as a result of "
+                  "macro arg concatenation.\n";
+      use_loc = expansion_loc;
+    } else if (GlobalSourceManager()->isMacroArgExpansion(use_loc)) {
+      VERRS(5) << "Use location is a macro arg expansion.\n";
+      use_loc = expansion_loc;
+    } else {
+      responsible = "definition";
+    }
+
+    VERRS(5) << "Attributing use to macro " << responsible << " location at "
+             << PrintableLoc(use_loc) << ".\n";
```

Comment:
What about something like
```c++
    VERRS(5) << "Attributing " << decl->getNameAsString() << " use "
             << "to macro " << responsible << " location at "
             << PrintableLoc(use_loc) << ".\n";
```
? I also tried `PrintableDecl(decl)` but it outputs
```
Attributing class Foo {
} use to macro expansion location at tests/cxx/macro_location.h:19:15.
```